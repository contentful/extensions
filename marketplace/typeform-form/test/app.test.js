'use strict';

const React = require('react');
const renderer = require('react-test-renderer');

const { mockComponent, flushPromises } = require('./helpers.js');

jest.mock('@contentful/forma-36-react-components', () => ({
  TextInput: mockComponent('text-input'),
  Select: mockComponent('select'),
  Option: mockComponent('option')
}));

jest.mock('../src/warning.js', () => ({
  Warning: mockComponent('warning')
}));

let mockFetchForms = () => Promise.resolve([]);

jest.mock('../src/fetch.js', () => ({
  fetchForms: (workspaceId, accessToken) => mockFetchForms(workspaceId, accessToken)
}));

const { App } = require('../src/app.js');

describe('App', function() {
  test('renders loading state', function() {
    const isRequired = false;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return undefined;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer
      .create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders error state', function() {
    mockFetchForms = () => Promise.reject(new Error("ain't working"));
    const isRequired = false;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return undefined;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('renders error state with the field being required', function() {
    mockFetchForms = () => Promise.reject(new Error("ain't working"));
    const isRequired = true;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return undefined;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('renders error state with previous value', function() {
    mockFetchForms = () => Promise.reject(new Error("ain't working"));
    const isRequired = false;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return 'https://cf-exts.typeform.com/to/ABCDEF';
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('renders with forms', function() {
    mockFetchForms = () =>
      Promise.resolve([
        {
          id: '123',
          title: 'First form',
          _links: { display: 'https://cf-exts.typeform.com/to/123' }
        },
        {
          id: '345',
          title: 'Second form',
          _links: { display: 'https://cf-exts.typeform.com/to/345' }
        },
        {
          id: '678',
          title: 'Third form',
          _links: { display: 'https://cf-exts.typeform.com/to/678' }
        }
      ]);
    const isRequired = false;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return undefined;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('renders with forms and being required', function() {
    mockFetchForms = () =>
      Promise.resolve([
        {
          id: '123',
          title: 'First form',
          _links: { display: 'https://cf-exts.typeform.com/to/123' }
        },
        {
          id: '345',
          title: 'Second form',
          _links: { display: 'https://cf-exts.typeform.com/to/345' }
        },
        {
          id: '678',
          title: 'Third form',
          _links: { display: 'https://cf-exts.typeform.com/to/678' }
        }
      ]);
    const isRequired = true;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return undefined;
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('renders with forms and previous value', function() {
    mockFetchForms = () =>
      Promise.resolve([
        {
          id: '123',
          title: 'First form',
          _links: { display: 'https://cf-exts.typeform.com/to/123' }
        },
        {
          id: '345',
          title: 'Second form',
          _links: { display: 'https://cf-exts.typeform.com/to/345' }
        },
        {
          id: '678',
          title: 'Third form',
          _links: { display: 'https://cf-exts.typeform.com/to/678' }
        }
      ]);
    const isRequired = true;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return 'https://cf-exts.typeform.com/to/123';
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  test('renders with forms and no longer existing previous value', function() {
    mockFetchForms = () =>
      Promise.resolve([
        {
          id: '123',
          title: 'First form',
          _links: { display: 'https://cf-exts.typeform.com/to/123' }
        },
        {
          id: '345',
          title: 'Second form',
          _links: { display: 'https://cf-exts.typeform.com/to/345' }
        },
        {
          id: '678',
          title: 'Third form',
          _links: { display: 'https://cf-exts.typeform.com/to/678' }
        }
      ]);
    const isRequired = true;
    const parameters = { accessToken: 'abc', workspaceId: '123' };
    const sdk = {
      field: {
        getValue() {
          return 'https://cf-exts.typeform.com/to/234';
        },
        onValueChanged() {},
        onIsDisabledChanged() {}
      }
    };

    const tree = renderer.create(<App isRequired={isRequired} parameters={parameters} sdk={sdk} />);

    return flushPromises().then(() => expect(tree.toJSON()).toMatchSnapshot());
  });
});
