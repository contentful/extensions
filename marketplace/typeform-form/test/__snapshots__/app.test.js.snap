// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders error state 1`] = `
Array [
  <div
    className="text-input"
    disabled={false}
    onChange={[Function]}
    required={false}
  />,
  <div
    className="warning"
    iconColor="negative"
  >
    Could not fetch forms from Typeform. Received the following error: ain't working
  </div>,
]
`;

exports[`App renders error state with previous value 1`] = `
Array [
  <div
    className="text-input"
    disabled={false}
    onChange={[Function]}
    required={false}
    value="https://cf-exts.typeform.com/to/ABCDEF"
  />,
  <div
    className="warning"
    iconColor="negative"
  >
    Could not fetch forms from Typeform. Received the following error: ain't working
  </div>,
]
`;

exports[`App renders error state with the field being required 1`] = `
Array [
  <div
    className="text-input"
    disabled={false}
    onChange={[Function]}
    required={true}
  />,
  <div
    className="warning"
    iconColor="negative"
  >
    Could not fetch forms from Typeform. Received the following error: ain't working
  </div>,
]
`;

exports[`App renders loading state 1`] = `null`;

exports[`App renders with forms 1`] = `
<div
  className="select"
  disabled={false}
  onChange={[Function]}
  required={false}
>
  <div
    className="option"
    value=""
  >
    Choose a typeform
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/123"
  >
    First form
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/345"
  >
    Second form
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/678"
  >
    Third form
  </div>
</div>
`;

exports[`App renders with forms and being required 1`] = `
<div
  className="select"
  disabled={false}
  onChange={[Function]}
  required={true}
>
  <div
    className="option"
    value=""
  >
    Choose a typeform
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/123"
  >
    First form
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/345"
  >
    Second form
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/678"
  >
    Third form
  </div>
</div>
`;

exports[`App renders with forms and no longer existing previous value 1`] = `
Array [
  <div
    className="warning"
  >
    The form you have selected in Contentful no longer exists in Typeform.
  </div>,
  <div
    className="select"
    disabled={false}
    onChange={[Function]}
    required={true}
    value="https://cf-exts.typeform.com/to/234"
  >
    <div
      className="option"
      value=""
    >
      Choose a typeform
    </div>
    <div
      className="option"
      value="https://cf-exts.typeform.com/to/234"
    >
      Previous value (no longer exists in Typeform)
    </div>
    <div
      className="option"
      value="https://cf-exts.typeform.com/to/123"
    >
      First form
    </div>
    <div
      className="option"
      value="https://cf-exts.typeform.com/to/345"
    >
      Second form
    </div>
    <div
      className="option"
      value="https://cf-exts.typeform.com/to/678"
    >
      Third form
    </div>
  </div>,
]
`;

exports[`App renders with forms and previous value 1`] = `
<div
  className="select"
  disabled={false}
  onChange={[Function]}
  required={true}
  value="https://cf-exts.typeform.com/to/123"
>
  <div
    className="option"
    value=""
  >
    Choose a typeform
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/123"
  >
    First form
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/345"
  >
    Second form
  </div>
  <div
    className="option"
    value="https://cf-exts.typeform.com/to/678"
  >
    Third form
  </div>
</div>
`;
